datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -68,11 +70,57 @@ impl Signature {
     }
 }
 
-/// Returns a [`Signature`] for applying `op` to arguments of type `lhs` and `rhs`
-fn signature(lhs: &DataType, op: &Operator, rhs: &DataType) -> Result<Signature> {
-    use arrow::datatypes::DataType::*;
-    use Operator::*;
-    match op {
+pub struct BinaryTypeCoercer<'a> {
```

Comment:
I like the idea of using this struct rather than a free function. If we are going to make it a `pub struct` I think it should have comments explaining what it does

Reply:
Added